File: app/Models/Prescription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Prescription extends Model
{
    protected $fillable = [
        'appointment_id', 'patient_id', 'doctor_id',
        'prescription_code', 'diagnosis', 'advice', 'follow_up_date', 'status', 'payment_status',
    ];

    protected $casts = [
        'follow_up_date' => 'date',
    ];

    public function appointment()
    {
        return $this->belongsTo(Appointment::class);
    }

    public function patient()
    {
        return $this->belongsTo(Patient::class);
    }

    public function doctor()
    {
        return $this->belongsTo(Doctor::class);
    }

    public function items()
    {
        return $this->hasMany(PrescriptionItem::class);
    }

    public function payments()
    {
        return $this->morphMany(Payment::class, 'payable');
    }

    public function getEstimatedTotalAttribute(): float
    {
        return (float) $this->items->sum(function ($i) {
            // 1) If linked to catalog, use that price.
            if ($i->medicine && $i->medicine->price > 0) {
                return $i->medicine->price;
            }
            // 2) Try matching the typed name against the catalog (case-insensitive).
            $byName = Medicine::whereRaw('LOWER(name) LIKE ?', [strtolower(trim($i->medicine_name)).'%'])
                ->value('price');
            if ($byName !== null && $byName > 0) {
                return (float) $byName;
            }
            // 3) Fallback so the pharmacist sees a starting figure they can edit.
            return 50.0;
        });
    }
}

File: mobile_app/lib/models/prescription.dart
import 'doctor.dart';
import 'patient.dart';

class Prescription {
  final int id;
  final int doctorId;
  final int patientId;
  final int? appointmentId;
  final String? diagnosis;
  final String? advice;
  final String? followUpDate;
  final DateTime createdAt;
  final String status;
  final String paymentStatus;
  final List<PrescriptionItem> items;
  final Doctor? doctor;
  final Patient? patient;

  Prescription({
    required this.id,
    required this.doctorId,
    required this.patientId,
    this.appointmentId,
    this.diagnosis,
    this.advice,
    this.followUpDate,
    required this.createdAt,
    this.status = 'issued',
    this.paymentStatus = 'unpaid',
    this.items = const [],
    this.doctor,
    this.patient,
  });

  factory Prescription.fromJson(Map<String, dynamic> json) => Prescription(
        id: json['id'] as int,
        doctorId: json['doctor_id'] as int,
        patientId: json['patient_id'] as int,
        appointmentId: json['appointment_id'] as int?,
        diagnosis: json['diagnosis'] as String?,
        advice: json['advice'] as String?,
        followUpDate: json['follow_up_date'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
        status: json['status'] as String? ?? 'issued',
        paymentStatus: json['payment_status'] as String? ?? 'unpaid',
        items: (json['items'] as List?)
                ?.map((e) => PrescriptionItem.fromJson(e))
                .toList() ??
            const [],
        doctor: json['doctor'] != null ? Doctor.fromJson(json['doctor']) : null,
        patient: json['patient'] != null ? Patient.fromJson(json['patient']) : null,
      );
}

class PrescriptionItem {
  final int id;
  final String medicineName;
  final String? dosage;
  final String? frequency;
  final String? duration;
  final String? instructions;

  PrescriptionItem({
    required this.id,
    required this.medicineName,
    this.dosage,
    this.frequency,
    this.duration,
    this.instructions,
  });

  factory PrescriptionItem.fromJson(Map<String, dynamic> json) => PrescriptionItem(
        id: json['id'] as int? ?? 0,
        medicineName: json['medicine_name'] as String? ?? '',
        dosage: json['dosage'] as String?,
        frequency: json['frequency'] as String?,
        duration: json['duration'] as String?,
        instructions: json['instructions'] as String?,
      );
}
